File: traverse.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;

namespace Traverse
{
  class Tree
  {
    private class TreeNode
    {
      public TreeNode parent;
      public TreeNode left;
      public TreeNode right;
      public Int64 value;

      public TreeNode(Int64 value_, TreeNode left_ = null, TreeNode right_ = null)
      {
        parent = null;
        left = left_;
        if (left != null)
          left.parent = this;
        right = right_;
        if (right != null)
          right.parent = this;
        value = value_;
      }
    }

    private TreeNode root;
    private TreeNode finger;

    private TreeNode BuildFullInternal(Int64 depth, Int64 value)
    {
      if (depth == 0)
      {
        return null;
      }

      TreeNode left = BuildFullInternal(depth - 1, value * 2);
      TreeNode right = BuildFullInternal(depth - 1, value * 2 + 1);
      return new TreeNode(value, left, right);
    }

    public Tree()
    { }

    public void BuildFull(int depth)
    {
      root = finger = BuildFullInternal(depth, 1);
    }

    public void FingerUp()
    {
      if (finger.parent == null)
      {
        return;
      }

      finger = finger.parent;
    }

    public void FingerLeft()
    {
      if (finger.left == null)
      {
        return;
      }

      finger = finger.left;
    }

    public void FingerRight()
    {
      if (finger.right == null)
      {
        return;
      }

      finger = finger.right;
    }

    public void FingerSet(Int64 value)
    {
      finger.value = value;
    }

    public void FingerReset()
    {
      finger = root;
    }
  }

  class Program
  {
    static void Convert(List<Int64> input, List<Int64> output)
    {
      Stack<Int64> stack = new Stack<Int64>();

      foreach (var v in input)
      {
        switch (v)
        {
          case 0:
            stack.Pop();
            break;
          case 1:
          case 2:
            stack.Push(v);
            break;
          default:
            output.AddRange(stack);
            output.Add(v);
            break;
        }
      }
    }

    static void InterpretFinger(Tree t, List<Int64> input)
    {
      foreach (var v in input)
      {
        switch (v)
        {
          case 0:
            t.FingerUp();
            break;
          case 1:
            t.FingerLeft();
            break;
          case 2:
            t.FingerRight();
            break;
          default:
            t.FingerSet(v);
            break;
        }
      }
    }

    static void InterpretRoot(Tree t, List<Int64> input)
    {
      foreach (var v in input)
      {
        switch (v)
        {
          case 0:
            break;
          case 1:
            t.FingerLeft();
            break;
          case 2:
            t.FingerRight();
            break;
          default:
            t.FingerSet(v);
            t.FingerReset();
            break;
        }
      }
    }

    static KeyValuePair<double, double> Analyze(List<double> input)
    {
      double mean = input.Average();

      double sumSq = 0.0;
      input.ForEach(x => sumSq += (x - mean) * (x - mean));
      double stddev = Math.Sqrt(sumSq / (input.Count - 1));

      return new KeyValuePair<double, double>(mean, stddev);
    }

    static void Main(string[] args)
    {
      int timeLimit = int.Parse(args[0]);

      List<Int64> inFinger = new List<Int64>();

      int depth = int.Parse(Console.ReadLine());

      {
        Int64 i;
        while (Int64.TryParse(Console.ReadLine(), out i))
        {
          inFinger.Add(i);
        }
      }

      List<Int64> inRoot = new List<Int64>();
      Convert(inFinger, inRoot);

      Tree t = new Tree();
      t.BuildFull(depth);

      Stopwatch local = new Stopwatch();
      Stopwatch total = new Stopwatch();

      {
        List<double> fingerResults = new List<double>();

        total.Restart();
        for (int i = 1;; i++)
        {
          local.Restart();
          for (int j = 0; j < i; j++)
          {
            t.FingerReset();
            InterpretFinger(t, inFinger);
          }
          local.Stop();
          fingerResults.Add((double)local.ElapsedMilliseconds / i);
          if (total.ElapsedMilliseconds > timeLimit)
            break;
        }

        var result = Analyze(fingerResults);
        Console.WriteLine("finger_mean = {0} ms", result.Key);
        Console.WriteLine("finger_stddev = {0} ms", result.Value);
        Console.Write("finger_results = ");
        foreach (var d in fingerResults)
        {
          Console.Write("{0},", d);
        }
        Console.WriteLine();
      }

      {
        List<double> rootResults = new List<double>();

        total.Restart();
        for (int i = 1;; i++)
        {
          local.Restart();
          for (int j = 0; j < i; j++)
          {
            t.FingerReset();
            InterpretRoot(t, inRoot);
          }
          local.Stop();
          rootResults.Add((double)local.ElapsedMilliseconds / i);
          if (total.ElapsedMilliseconds > timeLimit)
            break;
        }

        var result = Analyze(rootResults);
        Console.WriteLine("root_mean = {0} ms", result.Key);
        Console.WriteLine("root_stddev = {0} ms", result.Value);
        Console.Write("root_results = ");
        foreach (var d in rootResults)
        {
          Console.Write("{0},", d);
        }
        Console.WriteLine();
      }
    }
  }
}

File: traverse.cpp
#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <iostream>
#include <numeric>
#include <sstream>
#include <vector>

struct tree_t
{
private:
  struct node_t
  {
    node_t* parent;
    node_t* left;
    node_t* right;

    int64_t value;

    node_t(int64_t value_, node_t* left_ = nullptr, node_t* right_ = nullptr)
      : parent()
      , left(left_)
      , right(right_)
      , value(value_)
    {
      if (left)
        left->parent = this;
      if (right)
        right->parent = this;
    }
  };

  node_t* root;
  node_t* finger;

  void destroy(node_t* from)
  {
    if (!from)
      return;

    destroy(from->left);
    destroy(from->right);
    delete from;
  }

  node_t* build_full_internal(int depth, int64_t value)
  {
    if (depth == 0)
      return nullptr;

    node_t* left  = build_full_internal(depth - 1, value * 2);
    node_t* right = build_full_internal(depth - 1, value * 2 + 1);
    return new node_t(value, left, right);
  }

public:
  tree_t()
    : root()
    , finger()
  { }

  ~tree_t()
  {
    destroy(root);
  }

  void build_full(int depth)
  {
    destroy(root);
    root = finger = build_full_internal(depth, 1);
  }

  void finger_up()
  {
    if (!finger->parent)
      return;

    finger = finger->parent;
  }

  void finger_left()
  {
    if (!finger->left)
      return;

    finger = finger->left;
  }

  void finger_right()
  {
    if (!finger->right)
      return;

    finger = finger->right;
  }

  void finger_set(int64_t value)
  {
    finger->value = value;
  }

  void finger_reset()
  {
    finger = root;
  }
};

constexpr int64_t MOVE_UP = 0;
constexpr int64_t MOVE_LEFT = 1;
constexpr int64_t MOVE_RIGHT = 2;

// Convert finger movement instructions into node positions
// relative to the root.
template <typename ItIn, typename ItOut>
void convert(ItIn in_begin, ItIn in_end, ItOut out)
{
  std::vector<int64_t> stack;

  while (in_begin != in_end)
  {
    int64_t v = *in_begin++;
    switch (v)
    {
      case MOVE_UP:
        stack.pop_back();
        break;
      case MOVE_LEFT:
      case MOVE_RIGHT:
        stack.push_back(v);
        break;
      default:
        out = std::copy(stack.begin(), stack.end(), out);
        *out++ = v;
        break;
    }
  }
}

template <typename It>
void interpret_finger(tree_t& t, It begin, It end)
{
  while (begin != end)
  {
    int64_t v = *begin++;
    switch (v)
    {
      case MOVE_UP:
        t.finger_up();
        break;
      case MOVE_LEFT:
        t.finger_left();
        break;
      case MOVE_RIGHT:
        t.finger_right();
        break;
      default:
        t.finger_set(v);
        break;
    }
  }
}

template <typename It>
void interpret_root(tree_t& t, It begin, It end)
{
  while (begin != end)
  {
    int64_t v = *begin++;
    switch (v)
    {
      case MOVE_UP:
        break;
      case MOVE_LEFT:
        t.finger_left();
        break;
      case MOVE_RIGHT:
        t.finger_right();
        break;
      default:
        t.finger_set(v);
        t.finger_reset();
        break;
    }
  }
}

template <typename It>
std::pair<double, double> analyze(It begin, It end)
{
  int size = std::distance(begin, end);
  double mean = std::accumulate(begin, end, 0.0) / size;

  double sum_sq = 0.0;
  std::for_each(begin, end, [&] (double x) { sum_sq += (x - mean) * (x - mean); });
  double stddev = std::sqrt(sum_sq / (size - 1));

  return {mean, stddev};
}

int main(int argc, char** argv)
{
  using namespace std::chrono;

  int time_limit;

  {
    std::stringstream s(argv[1]);
    s >> time_limit;
  }

  std::vector<int64_t> in_finger;

  int depth;
  std::cin >> depth;

  {
    int64_t i;
    while (std::cin >> i)
      in_finger.push_back(i);
  }

  std::vector<int64_t> in_root;
  convert(in_finger.begin(), in_finger.end(), std::back_inserter(in_root));

  tree_t t;
  t.build_full(depth);

  {
    std::vector<double> finger_results;

    auto measurement_start = high_resolution_clock::now();
    for (int i = 1;; i++)
    {
      auto start = high_resolution_clock::now();
      for (int j = 0; j < i; j++)
      {
        t.finger_reset();
        interpret_finger(t, in_finger.begin(), in_finger.end());
      }
      auto end = high_resolution_clock::now();
      duration<double, std::milli> diff = end - start;
      finger_results.push_back(diff.count() / i);
      if (duration_cast<milliseconds>(end - measurement_start).count() > time_limit)
        break;
    }

    auto result = analyze(finger_results.begin(), finger_results.end());
    std::cout << "finger_mean = " << result.first << " ms\n";
    std::cout << "finger_stddev = " << result.second << " ms\n";
    std::cout << "finger_results = ";
    for (auto d : finger_results)
      std::cout << d << ",";
    std::cout << "\n";
  }

  {
    std::vector<double> root_results;

    auto measurement_start = high_resolution_clock::now();
    for (int i = 1;; i++)
    {
      auto start = high_resolution_clock::now();
      for (int j = 0; j < i; j++)
      {
        t.finger_reset();
        interpret_root(t, in_root.begin(), in_root.end());
      }
      auto end = high_resolution_clock::now();
      duration<double, std::milli> diff = end - start;
      root_results.push_back(diff.count() / i);
      if (duration_cast<milliseconds>(end - measurement_start).count() > time_limit)
        break;
    }

    auto result = analyze(root_results.begin(), root_results.end());
    std::cout << "root_mean = " << result.first << " ms\n";
    std::cout << "root_stddev = " << result.second << " ms\n";
    std::cout << "root_results = ";
    for (auto d : root_results)
      std::cout << d << ",";
    std::cout << "\n";
  }
}
